### ISSUE TITLE
Possible memory leak?

### ISSUE THREAD

**Issue by Daenyth** [HypothesisWorks/hypothesis-python#431]
I'm currently trying to debug a nasty memory leak in my test suite. Currently the only common factor in all of the suffering tests is that they use hypothesis, and the generated values get pretty large. Memory usage climbs rapidly until it exhausts available memory, and then the process more or less hangs. I've had this happen on both 16gb and 64gb systems.

**Comment by Daenyth** [comment]
It may also be celery-related but I'm not sure yet

**Comment by alexwlchan** [comment]
Are you able to share any of the tests that are causing the issue?

It’s definitely possible to exhaust memory with Hypothesis, if you inadvertently use its strategy values to compose large structures. Here’s a example of something I wrote recently that used all my available memory:

```python
from hypothesis import given, strategies as st

@given(st.integers())
def test_xs(n):
    assert 'a' in 'a' * abs(n)  # this string gets big. like, really big
```

(Not the exact code, but something to the same effect happened several levels deep in the code I was testing.)

**Comment by Daenyth** [comment]
I'm absolutely using it to compose very large structures. Often tuples of lists of large classes, sometimes with nested other classes.

The thing that seems odd to me is that it seems from my unscientific poking (watching htop with print statements all over the place) that memory seems to climb inside the test body more than in between test iterations. I'll try to look at it some more.

My main suspicion is when a test function calls some code that does a `celery_task.delay()`. I'm using the `ALWAYS_EAGER` celery setting so tasks just run in the calling thread immediately, but it's really hard to track down what exactly is consuming the memory. To my knowledge there's not a good tool like for the JVM that can tell you what class objects are taking up the most space

**Comment by Daenyth** [comment]
OK, watching it more closely now with even more print statements and some delays, I think this is definitely related to celery. I have a print right before a `slow_celery_task.delay()` and when output pauses there I see memory usage jump up by a gigabyte or more at each invocation

**Comment by DRMacIver** [comment]
There are definitely certain workloads that can cause Hypothesis to use a lot of memory, mostly due to some rather aggressive caching behaviour, so it's not impossible, but so far nobody has managed to hit them by accident, and they'll generally look like a slow upwards creep rather than a sudden jump.

My guess would be that it's more likely that either Hypothesis is triggering a problem in something else, or that you've got somewhere that you're e.g. passing an unbounded integer as a parameter to a function that tries to allocate O(n) memory (that latter one is the most common one I've seen)

**Comment by DRMacIver** [comment]
Oh BTW you can use [tracemalloc](https://docs.python.org/3/library/tracemalloc.html) to give you an idea of this if your code is Python 3 compatible.

**Comment by Daenyth** [comment]
It's unfortunately not completely py3k compatible yet, still a work in progress.

**Comment by DRMacIver** [comment]
There seems to be [a backport](https://pypi.python.org/pypi/pytracemalloc). No idea how good it is.

**Comment by Daenyth** [comment]
Actually I just tried running one of the affected test cases with python3 and this code path seems compatible so far. I'll explore a bit with tracemalloc

**Comment by Daenyth** [comment]
OK, I figured this out. It's a conflation of mock, unittest, hypothesis, and celery.

The short version is that someone added a unittest `setUp` method that mocked a logger method. The logger being mocked was set to run around our celery task `__call__` invocations, which is why I saw it happen with celery_task.delay(). The unittest setUp lifecycle is wider than hypothesis's runtime, so each test iteration leaked more mock data into the object.

Symptoms for anyone else experiencing this:
- Test suite leaks memory to system exhaustion
- You use unittest and hypothesis
- `tracemalloc` shows one mock in particular growing largely at each test iteration
- Sometimes you see a test fail with `RuntimeError: parent hierarchy exceeded 1000`
